### PR TITLE
[FLINK-21774][sql-client] Do not display column names when return set is emtpy in SQL Client

### PR DESCRIPTION
**Pull request by SteNicholas**
## What is the purpose of the change

*Currently, SQL Client displays the column names when the return set of `SHOW MODULES` is empty:*
```
SHOW MODULES;
+-------------+
| module name |
+-------------+
0 row in set
!ok
```
*The output of `SHOW MODULES` could be improved by simply omit the column names header.*

## Brief change log

  - *`CliClient` add `callShowResult` and `callShowCurrentResult` method to unify the client execution of `SHOW` and `SHOW` statement.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliClient.java [hunk 1]
```diff
@@ -781,6 +626,51 @@ private void callDdl(String ddl, String successMessage, String errorMessage) {
...
+                            .collect(Collectors.toList());
+        } catch (SqlExecutionException e) {
+            printExecutionException(e);
+            return;
+        }
+        if (results.isEmpty()) {
+            terminal.writer().println(CliStrings.messageInfo(CliStrings.MESSAGE_EMPTY).toAnsi());
+        } else {
+            if (isSorted) {
+                Collections.sort(results);
```
Comment (wuchong) [inline flink-table/flink-sql-client/src/main/java/org/apache/flink/table/client/cli/CliClient.java:651]: If we have sorted requirement, this should be sorted by TableEnvironment. The SQL Client should be a simple wrapper on TableEnv.

**Review by wuchong — CHANGES_REQUESTED** [review]
Thanks @SteNicholas , I think we can further improvment the column names displaying. Currently, may SHOW command doesn't display column names when there are results. Actualy, many mature database vendors will display column names (print like a table). 

I suggest all the SHOW commands should call `getResultAsTableauForm` instead which already print the column name headers.

**Comment by wuchong** [comment]
Besides, please rebase to the latest master branch.

**Comment by SteNicholas** [comment]
@wuchong , thanks for your detailed review. I have followed your suggestion to use `getResultAsTableauForm` for `SHOW` command. Please help to review again.

File: flink-table/flink-sql-client/src/test/resources/sql/function.q [hunk 2]
```diff
@@ -145,9 +170,14 @@ drop function if exists non_func;
...
-func11
-func3
-func4
++---------------+
+| function name |
++---------------+
+|         func4 |
+|        func11 |
+|         func3 |
++---------------+
```
Comment (wuchong) [inline flink-table/flink-sql-client/src/test/resources/sql/function.q:179]: We still need a sorted result, but the sort should be supported by TableEnvironment. Could you sort the result for `TableEnvironment#listFunctions` and `listUserDefinedFunctions`?

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/PrintUtils.java [hunk 3]
```diff
@@ -145,36 +145,30 @@ public static void printAsTableauForm(
...
-        }
-
-        final String rowTerm;
-        if (numRows > 1) {
-            rowTerm = "rows";
+            final String rowTerm = numRows > 1 ? "rows" : "row";
+            printWriter.println(numRows + " " + rowTerm + " in set");
         } else {
-            rowTerm = "row";
+            printWriter.println("Empty set");
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/PrintUtils.java:170]: nit: we can check empty set the beginning of this method and return.

**Review by wuchong — COMMENTED** [review]
I only have 2 minor comments.

**Comment by SteNicholas** [comment]
@wuchong , thanks for the minor comments. I have followed the suggestion for `listFunctions` and `listUserDefinedFunctions`. Please check again.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java [hunk 4]
```diff
@@ -628,12 +628,19 @@ public boolean dropTemporaryView(String path) {
...
     @Override
     public String[] listUserDefinedFunctions() {
-        return functionCatalog.getUserDefinedFunctions();
+        return sortFunctions(functionCatalog.getUserDefinedFunctions());
     }
 
     @Override
     public String[] listFunctions() {
-        return functionCatalog.getFunctions();
+        return sortFunctions(functionCatalog.getFunctions());
```
Comment (wuchong) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/TableEnvironmentImpl.java:636]: I think we can simply `Arrays.sort(functionCatalog.getFunctions())`. The returned `getFunctions` should never be null.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/PrintUtils.java [hunk 5]
```diff
@@ -116,65 +116,60 @@ public static void printAsTableauForm(
...
-                            printRowKind ? ROW_KIND_COLUMN : null);
-        } else {
-            final List<Row> rows = new ArrayList<>();
-            final List<String[]> content = new ArrayList<>();
-            content.add(columnNames);
-            while (it.hasNext()) {
-                Row row = it.next();
-                rows.add(row);
-                content.add(rowToString(row, nullColumn, printRowKind));
+        if (it.hasNext()) {
```
Comment (wuchong) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/utils/PrintUtils.java:119]: Personally, I don't like big if else block. Could you return the method when it is emtpy?

**Review by wuchong — APPROVED** [review]
Thanks for the updating. LGTM.

**Pull request merged**